### PR TITLE
DX-174: Kotlin Serialization Prototype

### PR DESCRIPTION
**Pull request by Funkatronics**
## Description
- Create a prototype serialization format for decoding Borsh encoded objects
- https://linear.app/metaplex/issue/dx-174

## Work Completed
- Borsh Format
  - this is an implementation of `BinaryFormat` that encodes and decodes data to the Borsh specification 
- Added new `AuctionHouse` serializable data class
- Added (temporary) serializable MetadataAccount classes to demonstrate how awesome the Serialization plugin is

## Testing
- see tests in `AuctionHouseUnitTests`: verified that we can successfully encode and decode from borsh coded data

### CODE REVIEW

File: lib/src/main/java/com/metaplex/lib/experimental/serialization/format/Borsh.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        val bytes = ByteArray(length)
+        byteBuffer.get(bytes)
+        return String(bytes, StandardCharsets.UTF_8).replace("\u0000", "")
+    }
+}
+
+class BorshEncoder : AbstractEncoder() {
+    // TODO: need to figure out how to compute the account discriminator,
+    //  for now we are just putting all zeros for the discriminator
+    private val bytes = mutableListOf<Byte>(0, 0, 0, 0, 0, 0, 0, 0)
```
Comment (Funkatronics) [inline lib/src/main/java/com/metaplex/lib/experimental/serialization/format/Borsh.kt:63]: the account discriminator is a Solana/Anchor thing, it is not part of the Borsh spec. Remove this and from the Borsh encoder/decoder and instead move it to its own abstraction that can be used on top of borsh

File: lib/src/main/java/com/metaplex/lib/experimental/serialization/serializers/rpc/RpcSerializers.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.buildClassSerialDescriptor
+import kotlinx.serialization.descriptors.element
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.*
+
+class JsonRpcSerializer2<D, ED>(val dataSerializer: KSerializer<D>,
+                                errorDataSerializer: KSerializer<ED>) : KSerializer<RpcResponse> {
```
Comment (Funkatronics) [inline lib/src/main/java/com/metaplex/lib/experimental/serialization/serializers/rpc/RpcSerializers.kt:21]: its annoying and potentially confusing that we are forced to define the returned error type, need to find a way to abstract this away and make it an optional advanced configuration if it's needed.

Alternatively we could simply ignore the additional error data that _might_ be present if we don't need to care about it.

File: lib/src/main/java/com/metaplex/lib/experimental/serialization/serializers/rpc/solana/SolanaRpcResponse.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
+/*
+ * SolanaRpcSerializers
+ * Metaplex
+ * 
+ * Created by Marco Martinez on 7/22/2022
+ * Copyright © Daxko, LLC 2022
```
Comment (Funkatronics) [inline lib/src/main/java/com/metaplex/lib/experimental/serialization/serializers/rpc/solana/SolanaRpcResponse.kt:6]: crap, my ide settings from last job. yikes, need to remove this

File: lib/src/main/java/com/metaplex/lib/experimental/serialization/serializers/rpc/RpcSerializers.kt (hunk 2 above)
Comment (Funkatronics) [reply lib/src/main/java/com/metaplex/lib/experimental/serialization/serializers/rpc/RpcSerializers.kt]: I think I will leave this as-is for now, and refactor it after I have used it a bit more (need get a sense of the devex to inform my decision here)

File: lib/src/main/java/com/metaplex/lib/experimental/serialization/serializers/solana/AnchorAccountSerializer.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.serializer
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+
+class AnchorAccountSerializer<T>(accountName: String, serializer: KSerializer<T>)
```
Comment (ajamaica) [inline lib/src/main/java/com/metaplex/lib/experimental/serialization/serializers/solana/AnchorAccountSerializer.kt:20]: Not all account responses are anchor accounts. How should we discriminate without specifying the serializer?

For example if we are deciding an nft metaplex account I doesn't contain that and the object @serializable doesn't provide enough information about the discriminator.

@anchorSerializable?

**Comment by ajamaica** [comment]
This is a great solution. Nail to the problem a super useful architecture for code generation

File: lib/src/test/java/com/metaplex/lib/serialization/rpc/solana/SolanaRpcResponseTests.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        canChangeSalePrice = false,
+        escrowPaymentBump = 0u,
+        hasAuctioneer = false,
+        auctioneerPdaBump = 0u
+    )
+
+    val json = Json {
+        ignoreUnknownKeys = true
+        serializersModule = SerializersModule {
+            contextual(ContextualDataSerializer(AnchorAccountSerializer<AuctionHouse>()))
```
Comment (ajamaica) [inline lib/src/test/java/com/metaplex/lib/serialization/rpc/solana/SolanaRpcResponseTests.kt:69]: This needs to be preconfigured? Can it be inferred from the type?

Comment (ajamaica) [reply lib/src/test/java/com/metaplex/lib/serialization/rpc/solana/SolanaRpcResponseTests.kt:69]: Or can we configure multiple?

File: lib/src/main/java/com/metaplex/lib/experimental/serialization/serializers/solana/AnchorAccountSerializer.kt (hunk 4 above)
Comment (Funkatronics) [reply lib/src/main/java/com/metaplex/lib/experimental/serialization/serializers/solana/AnchorAccountSerializer.kt:20]: Ha you are thinking exactly what I was thinking! I am trying to create a custom annotation, @AnchorAccount or something that would then tell the system to use that serializer.

File: lib/src/test/java/com/metaplex/lib/serialization/rpc/solana/SolanaRpcResponseTests.kt (hunk 5 above)
Comment (Funkatronics) [reply lib/src/test/java/com/metaplex/lib/serialization/rpc/solana/SolanaRpcResponseTests.kt:69]: in this case, yes it still needs to be pre configured. But this is something I am am trying to improve. Normally you don't need to configure like this, the system automatically figures out the correct serialize to use. This is a workaround needed for generic types, But I think i can improve it. 

And yes, multiple custom serializes can be configured.

**Comment by Funkatronics** [comment]
> This is a great solution. Nail the problem. a super useful architecture for code generation too.
> 
> I think the is a problem. It may suffer the same problem as moshi since the serializer must be configured and May require knowledge of an object. If we can deduct the serializer from the class it would solve the problem (it's how iOS works)

Each serializer provides a SerialDescriptor object which is a format agnostic representation of the class and its properties. This descriptor could easily be generated from the IDL json or something similar. So we definitely could deduct the serializer at runtime, or do some code gen.

**Pull request merged**